slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala

```diff
@@ -47,41 +92,76 @@ object SloMonitor {
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 210 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_210,
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 240 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_240,
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 270 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_270,
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 300 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_300,
+       |  COUNT(*) AS total_deliveries
+       |FROM notification_received_${stage.toLowerCase()}
+       |WHERE notificationid = '$notificationId'
+       |AND $partitionDate
+       |GROUP BY platform
+     """.stripMargin
   }
 
   def handleMessage(event: SQSEvent): Unit = {
     logger.info(s"Running SLO monitor")
     val record = event.getRecords.get(0) // Batch size is defined as 1 in cdk
     val notificationId = record.getBody
     val sentTime: LocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(record.getAttributes.get("SentTimestamp").toLong), ZoneOffset.UTC)
-    // The number of expected deliveries is hardcoded for the purposes of this prototype, but we could:
-    // a) Send this as part of the SQS message (from the notifications app)
-    // b) Obtain it via the DynamoDB Report table
-    // c) Something else...
-    val expectedDeliveries: Int = 100 // FIXME
     val query = generateQueryString(notificationId, sentTime)
     logger.info(s"Starting query: $query")
-    val result = Athena.startQuery(Query("notifications", query, s"s3://aws-mobile-event-logs-${this.stage.toLowerCase()}/athena/slo-monitoring"))
-      .flatMap(Athena.fetchQueryResponse(_, rows => rows.head.head)
-        .map(actualDeliveries => {
-          // val percentage = calculatePercentage(actualDeliveries.toDouble, expectedDeliveries.toDouble)
-          // In the future we could log structured data (for visualisation in Kibana) & send this as a CloudWatch metric here
-
-          // For the purposes of validating the approach we decided to log only the actual deliveries
-          // We will compare the actual deliveries logged against the data we see in bigquery as a way to validate accuracy of the solution/setup
-          logger.info(Map(
-            "notificationId" -> notificationId,
-            "deliveriesWithin2mins" -> actualDeliveries
-          ),
-          s"Notifications delivered within 120 seconds was $actualDeliveries")
-        })
-      )
+    val result = Athena.startQuery(Query("notifications", query, s"s3://aws-mobile-event-logs-${this.stage.toLowerCase()}/athena/slo-monitoring")).flatMap {
+      Athena.fetchQueryResponse(_, rows => {
+        val androidDeliveries = rows.head
```

Comment:
The response data looks like this:

```
"platform","less_than_30","less_than_60","less_than_90","less_than_120","less_than_150","less_than_180","less_than_210","less_than_240","less_than_270","less_than_300","total_deliveries"
"android","424","432","436","437","439","441","441","441","441","441","441"
"ios","701","702","703","704","704","704","704","704","704","704","704"
```

It looks like the `GROUP BY` is sorting alphabetically by default but perhaps we should add some validation or something here to ensure that the query results look roughly how we expect (e.g. there should be X rows, Y columns, one row starting with `android` and one with `ios` etc.)

Reply:
Great idea to separate this by platform! I guess we could use metric math in the console to get an aggregated view of the 90in2 (if we ever wanted to do this)?